### ISSUE TITLE
What does random forest mode do exactly?

### ISSUE THREAD

**Issue by mfluegge** [microsoft/LightGBM#2340]
Hey,

I'm looking for a python random forest implementation that
* can deal with missing values natively
* can find splits on categorical variables
* is fast

LightGBM seems like the way to go, since xgb doesn't support categorical variables. 
However, I think an important part in random forest is that you only consider a small subset of features on every split. Does the `rf` mode achieve this? I saw a feature + pull request about feature subsample on node-level (thats not merged yet), so I guess lightgbm doesn't support this yet? Is all that `rf` does then to disable boosting?

Also [the FAQ](https://lightgbm.readthedocs.io/en/latest/FAQ.html#i-tried-to-use-random-forest-mode-and-lightgbm-crashes) specifies that I should use a `bagging_fraction` different from 1. This doesn't really make sense to me. As I understand it, it is perfectly normal for random forests to use bootstramp samples that have the **the same number of samples** as the original dataset. Because one samples with replacement, however, not every sample from the original dataset is included in each boostrap sample. But still, why would `bagging_fraction = 1` be a problem?

**Comment by guolinke** [comment]
@mfluegge 
1. rf mode support sub-features. But currently, we only support the sub-feature at tree level, not the node level. I think the original rf also uses the sub-features at tree level.
2. we don't support the sample with replacement, therefore, bagging_fraction=1 does not make sense.

the rf mode is actually a hack of boosting mode. rf mode doesn't boost, and always uses the gradients of the first iteration to train. And we recommend the `gbdt` mode, for it can consistently outperform `rf`, with reasonable parameters.

**Comment by mfluegge** [comment]
Hey @guolinke, 

thanks for your response !

1. Ok, I will have to check how splitting on tree-level impacts the behaviour of a random forest. I'm mostly referring to [Leo Breiman's influential paper](https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf) which specifies random forest as bagged trees with random feature subsets selected at each split.

2. Alright, I understand. I guess I should set the bagging fraction to ~0.63 then, since a bootstrap sample (with replacement) with same size as training set contains on average ~63% of the training set samples.

I'm actually in a position where models are trained automatically with varying amounts of samples, features, feature scales and nans. If I were to manually set hyperparameters for each problem, then I'd definitely prefer boosting, but for my use-case I think I'd prefer the robustness of random forests to all these variations. 

Thanks again!
